src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/TargetFrameworkProjectConfigurationDimensionProvider.cs

```diff
@@ -0,0 +1,84 @@
+    [Export(typeof(IProjectConfigurationDimensionsProvider))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class TargetFrameworkProjectConfigurationDimensionProvider : IProjectConfigurationDimensionsProvider
+    {
+        private const string TargetFrameworkPropertyName = "TargetFramework";
+        private const string DefaultTargetFrameworkValue = "netcoreapp1.0";
+        private readonly IProjectLockService _projectLockService;
+        
+        [ImportingConstructor]
+        public TargetFrameworkProjectConfigurationDimensionProvider(IProjectLockService projectLockService)
+        {
+            Requires.NotNull(projectLockService, nameof(projectLockService));
+
+            _projectLockService = projectLockService;
+        }
+
+        public async Task<IEnumerable<KeyValuePair<string, string>>> GetDefaultValuesForDimensionsAsync(UnconfiguredProject project)
+        {
+            // First target framework is the default one.
+            var targetFrameworks = await GetOrderedTargetFrameworksAsync(project).ConfigureAwait(false);
+            return ImmutableArray.Create(new KeyValuePair<string, string>(TargetFrameworkPropertyName, targetFrameworks.FirstOrDefault()));
+        }
+
+        public async Task<IEnumerable<KeyValuePair<string, IEnumerable<string>>>> GetProjectConfigurationDimensionsAsync(UnconfiguredProject project)
+        {
+            var targetFrameworks = await GetOrderedTargetFrameworksAsync(project).ConfigureAwait(false);
+            if (targetFrameworks.IsEmpty)
+            {
+                return ImmutableArray<KeyValuePair<string, IEnumerable<string>>>.Empty;
+            }
+
+            return ImmutableArray.Create(new KeyValuePair<string, IEnumerable<string>>(TargetFrameworkPropertyName, targetFrameworks));
+        }
+
+        private async Task<ImmutableArray<string>> GetOrderedTargetFrameworksAsync(UnconfiguredProject project)
+        {
+            Requires.NotNull(project, nameof(project));
+
+            using (var access = await _projectLockService.ReadLockAsync())
+            {
+                // Read the "TargetFrameworks" properties from msbuild project evaluation.
+                var configuredProject = await project.GetSuggestedConfiguredProjectAsync().ConfigureAwait(false);
+                var msbuildProject = await access.GetProjectAsync(configuredProject).ConfigureAwait(false);
+                var targetFrameworksStrings = msbuildProject.Properties.Where(p => p.Name.Equals(ConfigurationGeneral.TargetFrameworksProperty, StringComparison.OrdinalIgnoreCase));
+
+                if (!targetFrameworksStrings.Any())
+                {
+                    return ImmutableArray<string>.Empty;
+                }
+
+                // We need to ensure that we return the target frameworks in specified order.
+                var targetFrameworks = ImmutableArray.CreateBuilder<string>();
+                foreach (var frameworksString in targetFrameworksStrings)
```

Comment:
There should be just one value in targetFrameworksStrings right? I don't know what the msbuild model returns if you have something like:

```
<TargetFrameworks>net451</TargetFrameworks>
<TargetFrameworks>net452</TargetFrameworks>
```

Since in the evaluation model it will only have net452 it should return only that entry in which case we don't have to loop here (instead do a .Single())

If it does return both then we should maybe do .Last() - otherwise we are including net451 as well which is incorrect.